### PR TITLE
Rework arithmetic operators.

### PR DESCRIPTION
**Pull request by fengttt**
Plus first.  If it looks ok, I will work out the rest.

**What type of PR is this?**

- [ ] API-change
- [ ] BUG
- [x] Improvement
- [ ] Documentation
- [ ] Feature
- [x] Test and CI
- [x] Code Refactoring

**Which issue(s) this PR fixes:**

issue #

**What this PR does / why we need it:**

Not Available

**Special notes for your reviewer:**

Not Available

**Additional documentation (e.g. design docs, usage docs, etc.):**

Not Available

### CODE REVIEW

**Review by ouyuanning — APPROVED** [review]

**Review by qingxinhome — APPROVED** [review]

**Review by daviszhen — APPROVED** [review]

File: pkg/vectorize/add/add.go [hunk 1]
```diff
@@ -17,310 +17,120 @@ package add
...
-func NumericAddScalarSigned[T constraints.Signed](x T, ys, rs []T) []T {
-	for i, y := range ys {
-		rs[i] = x + y
-		if x > 0 && y > 0 && rs[i] <= 0 {
-			panic(moerr.NewError(moerr.OUT_OF_RANGE, "int add overflow"))
+func NumericAddSigned[T constraints.Signed](xs, ys, rs *vector.Vector) error {
+	xt, yt, rt := vector.MustTCols[T](xs), vector.MustTCols[T](ys), vector.MustTCols[T](rs)
+	if xs.IsScalar() {
+		for i, y := range yt {
+			if !nulls.Contains(rs.Nsp, uint64(i)) {
```
[the remark this answers is not in the document]
Comment (fengttt) [reply pkg/vectorize/add/add.go:37]: Yes, maybe, but you have to be correct first ...

File: cgo/decimal.c [hunk 2]
```diff
@@ -582,4 +582,92 @@ int32_t Decimal128_DivInt64(int64_t *r, int64_t *a, int64_t b)
 	return Decimal128_Div(r, a, (int64_t *) &db);
 }
 
-
+int32_t Decimal64_Add_SV(int64_t *r, int64_t *a, int64_t *b, uint64_t n, uint64_t *nulls) {
```
[the remark this answers is not in the document]
Comment (fengttt) [reply cgo/decimal.c:585]: Yes, scalar vector.    I am thinking move this as a flag into C code.    So C code with looks like 
Decimal64_VectorAdd(...,  ScalarFlag);

ScalarFlag is 0 -> no scalar, 1 means left scalar 2 means right and 3 means both.   Not sure if I should.

File: cgo/decimal.c [hunk 3]
```diff
@@ -582,4 +582,92 @@ int32_t Decimal128_DivInt64(int64_t *r, int64_t *a, int64_t b)
...
+			int ret = Decimal128_Add(r+i+i, a, b+i+i);
+			if (ret != RC_SUCCESS) {
+				return ret;
+			}
+		}
+		return RC_SUCCESS;
+	}
+}
+	
+int32_t Decimal128_Add_VV(int64_t *r, int64_t *a, int64_t *b, uint64_t n, uint64_t *nulls) {
```
[the remark this answers is not in the document]
Comment (fengttt) [reply cgo/decimal.c:651]: Subtle.    Same signature, but these int64_t * points to different things.

Should I write a macro ...   Let me think, maybe next check in I will do.

File: pkg/vectorize/add/add.go [hunk 4]
```diff
@@ -17,310 +17,120 @@ package add
...
-// this is the slowest overflow check
-func NumericAddSigned[T constraints.Signed](xs, ys, rs []T) []T {
-	for i, x := range xs {
-		rs[i] = x + ys[i]
-		if x > 0 && ys[i] > 0 && rs[i] <= 0 {
-			panic(moerr.NewError(moerr.OUT_OF_RANGE, "int add overflow"))
-		}
-	}
-	return rs
+func signedOverflow[T constraints.Signed](a, b, c T) bool {
```
[the remark this answers is not in the document]
Comment (fengttt) [reply pkg/vectorize/add/add.go:25]: The overflow check mentioned there essentially is the same check here.    But the INT_MAX depends on T, so it is tiny bit harder.   

I have an older try, that run int addition in CGO, maybe we should activate that.   Next.

**Pull request merged**